Nextflow docs: BLAST example (#218)
latchbio/latch: docs/source/tutorials/nextflow.md

> To follow along, open the Jupyter Notebook provided inside the GitHub repo (`blast_nextflow_example.ipynb`).

## Step 2: Install dependencies for the Nextflow Pipeline

To successfully run Nextflow BLAST pipeline inside the Pod, your Pod environment needs:

* Java 8
* Nextflow (version 20.07.x or higher)

In this tutorial, we will use `conda` to manage the pipeline dependencies. We need to additionally install:

* Micromamba (a faster alternative to Anaconda)
* BLAST

Let's walk through our Jupyter Notebook to see how these dependencies are installed!

> Note: All the commands below assume that you are in a Linux environment. Please ensure that you are inside a Latch Pod or an alternative Linux environment before proceeding.

### Update system dependencies

First, let's download and update existing dependencies on our system:

```console
apt-get update -y && apt-get install -y curl unzip git
```

### Install Java 8

Java 8 is required to run Nextflow. You can install the headless version of Java 8 like so:

```console
apt-get install -y default-jre-headless
```

### Install Nextflow

```console
curl -s https://get.nextflow.io | bash && \
    mv nextflow /usr/bin/ && \

>     chmod 777 /usr/bin/nextflow 

Reviewer: ```suggestion
    chmod u+x /usr/bin/nextflow 
```

```console
# Install micromamba
! export CONDA_DIR=/opt/conda
! export MAMBA_ROOT_PREFIX=/opt/conda
! export PATH=$CONDA_DIR/bin:$PATH

! apt-get update && apt-get install -y wget bzip2 \
    && wget -qO-  https://micro.mamba.pm/api/micromamba/linux-64/latest | tar -xvj bin/micromamba \
    && touch /root/.bashrc \
    && ./bin/micromamba shell init -s bash -p /opt/conda  \
    && grep -v '[ -z "\$PS1" ] && return' /root/.bashrc  > /opt/conda/bashrc
    && apt-get clean autoremove --yes \
    && rm -rf /var/lib/{apt,dpkg,cache,log}
```

The above set of commands install Micromamba under the `/root/bin` directory.

### Create a Conda environment for the BLAST Nextflow Pipeline

We can use YAML files to manage conda dependencies. Inspecting the `blast-nf` folder, there is a `conda.yml` file that specifies `blast`, which is the only dependency required for this pipeline.

```yml
# blast-nf/conda.yml
name: blast-nf
channels:
  - defaults
  - bioconda
  - conda-forge
dependencies:
  - blast
```

You can create an environment called `blast-nf` using Micromamba like so:

```console
/root/bin/micromamba create -f /root/blast-nextflow-latch/blast-nf/conda.yml -y
```

> You can verify the dependencies installation by listing our the environment:

Reviewer 2: ```suggestion
You can verify the dependencies installation by listing the environments:
```

3. A `version` file to semantically name the workflow version

### Define your Dockerfile to install dependencies

As the Latch workflow will be executed on a fresh machine on the Latch platform, we have to define a Dockerfile with the necessary dependencies for BLAST to run.

To do so, we can copy paste previous commands used to set up our environment:

![Nextflow Dockerfile](../assets/nextflow-example/nextflow-dockerfile.png)

* **Line 1**: is the [Latch base image](https://github.com/latchbio/latch-base), which is used to configure libraries required for consistent task behaviour.
* **Line 3-4**: downloads and installs the updates for each outdated package and dependency on the machine that executes the workflow. `curl` and `unzip` are also installed.
* **Line 7**: installs the Java runtime environment, which is required to run Nextflow.
* **Line 8-10**: installs Nextflow and moves the binary to `/usr/bin`.
* **Line 13-25**: is a series of commands to install Micromamba.
* **Line 27**: copies the BLAST Nextflow pipeline code to the task execution environment. The `/root/blast-nf` is that path at which the NF code is stored in the machine that executes the task on Latch.

> * **Line 30**: uses Micromamba to install the dependencies as spefieid in `/root/blast-nf/conda.yml`.

Reviewer 2: ```suggestion
* **Line 30**: uses Micromamba to install the dependencies as specified in `/root/blast-nf/conda.yml`.
```

* **Line 35-39**: are already provided in the boilerplate Dockerfile and are needed to ensure your build envrionment works correctly with Latch.

That's it! You've successfully defined your Dockerfile.

To test whether the Dockerfile builds the correct environment, first register your workflow to Latch:

```console
latch register -r .
```

Then, open a remote debugging session:

```console
latch develop .
```

> Enter an interative shell:

Reviewer 2: ```suggestion
Enter an interactive shell:
```

Now you have successfully defined a Latch task with custom compute resources to execute the BLAST Nextflow pipeline on Latch!

### Calling a Latch task inside a Latch workflow

Since this is a single task workflow, you can simply call the task inside the workflow and return its results like so:

```python
@workflow(metadata)
def blast_wf(
    query: LatchFile, db: LatchDir, out: str
) -> LatchFile:
    ...
    return blast_task(query=query, db=db, out=out)
```

### Defining Workflow GUI

To expose workflow parameters to a user-friendly workflow GUI, you can use the `LatchMetadata` object. An important point to note is that all workflow arguments need to be added to the `parameters` key of LatchMetadata for them to display on the GUI. For an exhaustive list of how workflow argument and their Python types map to the front-end interface, visit [Customizing Your Interface](../basics/customizing_interface.md)

```python
"""The metadata included here will be injected into your interface."""
metadata = LatchMetadata(
    display_name="Example: Wrapping a Nextflow BLAST Pipeline in Latch SDK",
    documentation="your-docs.dev",
    author=LatchAuthor(

>         name="John von Neumann",

Reviewer: ?

>         email="hungarianpapi4@gmail.com",

Reviewer: ???
